File: src/main/java/me/alegian/thavma/impl/Thavma.java
package me.alegian.thavma.impl;

import com.mojang.logging.LogUtils;
import me.alegian.thavma.impl.client.event.T7ClientGameEventsKt;
import me.alegian.thavma.impl.client.event.T7ClientModEventsKt;
import me.alegian.thavma.impl.common.event.T7CommonGameEventsKt;
import me.alegian.thavma.impl.common.event.T7CommonModEventsKt;
import me.alegian.thavma.impl.init.registries.deferred.*;
import net.minecraft.resources.ResourceLocation;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.fml.common.Mod;
import org.slf4j.Logger;

// must match value in META-INF/mods.toml
@Mod(Thavma.MODID)
public class Thavma {
  public static final String MODID = "thavma";
  public static final Logger LOGGER = LogUtils.getLogger();

  // FML will recognize some parameter types like IEventBus or ModContainer and pass them in automatically.
  public Thavma(IEventBus modEventBus) {
    T7ArmorMaterials.INSTANCE.getREGISTRAR().register(modEventBus);
    T7Blocks.REGISTRAR.register(modEventBus);
    T7Items.REGISTRAR.register(modEventBus);
    T7BlockEntities.REGISTRAR.register(modEventBus);
    T7EntityTypes.REGISTRAR.register(modEventBus);
    T7Attachments.REGISTRAR.register(modEventBus);
    T7DataComponents.REGISTRAR.register(modEventBus);
    T7ParticleTypes.REGISTRAR.register(modEventBus);
    T7CreativeModeTabs.REGISTRAR.register(modEventBus);
    T7TrunkPlacerTypes.REGISTRAR.register(modEventBus);
    T7MenuTypes.REGISTRAR.register(modEventBus);
    T7RecipeTypes.REGISTRAR.register(modEventBus);
    T7RecipeSerializers.REGISTRAR.register(modEventBus);
    WandCoreMaterials.INSTANCE.getREGISTRAR().register(modEventBus);
    WandHandleMaterials.INSTANCE.getREGISTRAR().register(modEventBus);
    Aspects.REGISTRAR.register(modEventBus);
    Researches.INSTANCE.getREGISTRAR().register(modEventBus);
    T7Attributes.INSTANCE.getREGISTRAR().register(modEventBus);
    T7GlobalLootModifierSerializers.INSTANCE.getREGISTRAR().register(modEventBus);

    T7CommonModEventsKt.registerCommonModEvents();
    T7ClientModEventsKt.registerClientModEvents();
    T7CommonGameEventsKt.registerCommonGameEvents();
    T7ClientGameEventsKt.registerClientGameEvents();
  }

  public static ResourceLocation rl(String path) {
    return ResourceLocation.fromNamespaceAndPath(Thavma.MODID, path);
  }
}

File: src/main/java/me/alegian/thavma/impl/Thavma.kt
package me.alegian.thavma.impl

import com.mojang.logging.LogUtils
import me.alegian.thavma.impl.client.event.registerClientGameEvents
import me.alegian.thavma.impl.client.event.registerClientModEvents
import me.alegian.thavma.impl.common.event.registerCommonGameEvents
import me.alegian.thavma.impl.common.event.registerCommonModEvents
import me.alegian.thavma.impl.init.registries.deferred.*
import net.minecraft.resources.ResourceLocation
import net.neoforged.fml.common.Mod
import org.slf4j.Logger
import thedarkcolour.kotlinforforge.neoforge.forge.MOD_BUS as KFF_MOD_BUS

// must match value in META-INF/mods.toml
@Mod(Thavma.MODID)
object Thavma {
    const val MODID: String = "thavma"
    val LOGGER: Logger = LogUtils.getLogger()
    
    // FML will recognize some parameter types like IEventBus or ModContainer and pass them in automatically.
    init {
        T7ArmorMaterials.REGISTRAR.register(KFF_MOD_BUS)
        T7Blocks.REGISTRAR.register(KFF_MOD_BUS)
        T7Items.REGISTRAR.register(KFF_MOD_BUS)
        T7BlockEntities.REGISTRAR.register(KFF_MOD_BUS)
        T7EntityTypes.REGISTRAR.register(KFF_MOD_BUS)
        T7Attachments.REGISTRAR.register(KFF_MOD_BUS)
        T7DataComponents.REGISTRAR.register(KFF_MOD_BUS)
        T7ParticleTypes.REGISTRAR.register(KFF_MOD_BUS)
        T7CreativeModeTabs.REGISTRAR.register(KFF_MOD_BUS)
        T7TrunkPlacerTypes.REGISTRAR.register(KFF_MOD_BUS)
        T7MenuTypes.REGISTRAR.register(KFF_MOD_BUS)
        T7RecipeTypes.REGISTRAR.register(KFF_MOD_BUS)
        T7RecipeSerializers.REGISTRAR.register(KFF_MOD_BUS)
        WandCoreMaterials.REGISTRAR.register(KFF_MOD_BUS)
        WandHandleMaterials.REGISTRAR.register(KFF_MOD_BUS)
        Aspects.REGISTRAR.register(KFF_MOD_BUS)
        Researches.REGISTRAR.register(KFF_MOD_BUS)
        T7Attributes.REGISTRAR.register(KFF_MOD_BUS)
        T7GlobalLootModifierSerializers.REGISTRAR.register(KFF_MOD_BUS)

        registerCommonModEvents()
        registerClientModEvents()
        registerCommonGameEvents()
        registerClientGameEvents()
    }

    fun rl(path: String): ResourceLocation {
        return ResourceLocation.fromNamespaceAndPath(MODID, path)
    }
}
